### PR TITLE
Deduplicate more sized errors on call exprs

### PR DESCRIPTION
**Pull request by estebank**
Change the implicit `Sized` `Obligation` `Span` for call expressions to include the whole expression. This aids the existing deduplication machinery to reduce the number of errors caused by a single unsized expression.

### CODE REVIEW

File: tests/ui/parser/recover/recover-fn-trait-from-fn-kw.stderr [hunk 1]
```diff
@@ -30,8 +30,9 @@ error[E0631]: type mismatch in closure arguments
   --> $DIR/recover-fn-trait-from-fn-kw.rs:10:5
    |
 LL |     foo2(|_: ()| {});
-   |     ^^^^ ------- found signature defined here
-   |     |
+   |     ^^^^^-------^^^^
```
Comment (nnethercote) [inline tests/ui/parser/recover/recover-fn-trait-from-fn-kw.stderr:33]: This one seems weird?

File: tests/ui/inference/erase-type-params-in-label.stderr [hunk 2]
```diff
@@ -2,7 +2,7 @@ error[E0283]: type annotations needed for `Foo<i32, &str, W, Z>`
   --> $DIR/erase-type-params-in-label.rs:2:9
    |
 LL |     let foo = foo(1, "");
-   |         ^^^   --- type must be known at this point
+   |         ^^^   ---------- type must be known at this point
```
Comment (nnethercote) [inline tests/ui/inference/erase-type-params-in-label.stderr:5]: This is an orthogonal issue but seems worth mentioning: it took me a while to realize that these `---` sequences are pointing to the code above them. It's much less obvious than the `^^^` sequences. I just interpreted it as text to lead the reader's eye towards the `^^^`.

File: tests/ui/async-await/async-is-unwindsafe.stderr [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
+   |  |_____|
+   | ||
+LL | ||
+LL | ||         use std::ptr::null;
+LL | ||         use std::task::{Context, RawWaker, RawWakerVTable, Waker};
+...  ||
+LL | ||         drop(cx_ref);
+LL | ||     });
+   | ||_____-^ `&mut Context<'_>` may not be safely transferred across an unwind boundary
+   |  |_____|
```
Comment (nnethercote) [inline tests/ui/async-await/async-is-unwindsafe.stderr:15]: Orthogonal to this PR, but these arrows are really hard to follow.

File: tests/ui/anonymous-higher-ranked-lifetime.stderr [hunk 4]
```diff
@@ -2,8 +2,9 @@ error[E0631]: type mismatch in closure arguments
   --> $DIR/anonymous-higher-ranked-lifetime.rs:2:5
    |
 LL |     f1(|_: (), _: ()| {});
-   |     ^^ -------------- found signature defined here
-   |     |
+   |     ^^^--------------^^^^
```
Comment (nnethercote) [inline tests/ui/anonymous-higher-ranked-lifetime.stderr:5]: There are lots of new occurrences of `^^^---^^^` now. I guess the `^^^` highlighting and the `---` highlighting overlap, and the `---` wins?

Also, it's weird that the `f1` and the `{}` are both pointed to by the `^` markers, they don't really seem connected?

**Review by nnethercote — COMMENTED** [review]
This kind of diagnostic tweaking isn't my forte. But the deduplication is definitely nice. I find some of the `^^^---^^^` markers very hard to read, as mentioned below.

r=me assuming reasonable responses to the concerns below, and if you don't want a second opinion.

File: tests/ui/inference/erase-type-params-in-label.stderr (hunk 2 above)
Comment (nnethercote) [reply tests/ui/inference/erase-type-params-in-label.stderr:5]: It's likely I've been overlooking the spans demarcated by these `---` markers for  years.

Comment (estebank) [reply tests/ui/inference/erase-type-params-in-label.stderr:5]: 😮 

I'm really surprised to hear! They mark the secondary spans, usually giving you extra context. *At least*, even if you didn't realize that they have the benefit of bringing interesting code into view (if it didn't happen to be in the lines of the primary span).

File: tests/ui/async-await/async-is-unwindsafe.stderr (hunk 3 above)
Comment (estebank) [reply tests/ui/async-await/async-is-unwindsafe.stderr:15]: I'd like to clean up the rendering of overlapping multiline spans, but span labels make them even harder to deal with than usual.

File: tests/ui/anonymous-higher-ranked-lifetime.stderr (hunk 4 above)
Comment (estebank) [reply tests/ui/anonymous-higher-ranked-lifetime.stderr:5]: There's some logic so when primary and secondary spans overlap some bit of information is shown: the largest one is printed "first", and the then each shorter one is printed in order of length, so that at least a bit of each is shown. This is because if, for example in this case, there is a span with a label pointing within a larger one and we *weren't* doing that, we'd have no way of knowing where the label span ended within the larger span. That's how you end up with this kind of rendering. IIRC, we also tried some tricks with colors to aid here, but don't think they panned out so we didn't land them.

In this case the primary span marks from the `fq` all the way to the closing `)` (including the `{}`), it's just "visually interrupted" by the span pointing at the closure header.

**Comment by estebank** [comment]
FWIW, deduplication could also have been accomplished by going the other direction, using the path span for the obligation cause of the *other* cases, either way they get picked up by the "don't show the same trait bound failure twice" logic. That might have been nice, looking at it with fresh eyes, just don't know how much extra effort that would be.

File: tests/ui/anonymous-higher-ranked-lifetime.stderr (hunk 4 above)
Comment (nnethercote) [reply tests/ui/anonymous-higher-ranked-lifetime.stderr:5]: Makes sense. Not for this PR, but I wonder if changing this:
```
LL |     f1(|_: (), _: ()| {});
   |     ^^^--------------^^^^
   |     |  |
   |     |  found signature defined here
   |     expected due to this
```
to this:
```
LL |     f1(|_: (), _: ()| {});
   |     ^^^^^^^^^^^^^^^^^^^^^
   |     |  --------------
   |     |  |
   |     |  found signature defined here
   |     expected due to this
```
would help.

W.r.t. my lack of understanding of `---` marking the secondary span, I think it's a lot clearer that `---` is marking code when the secondary span label is underneath the `---` and there's a `|` leading to it, instead of the label being to the right on the same line.

File: tests/ui/inference/erase-type-params-in-label.stderr (hunk 2 above)
Comment (nnethercote) [reply tests/ui/inference/erase-type-params-in-label.stderr:5]: I'm sure people have suggested using non-ASCII chars for this sort of thing, and presumably there's a reason why it's still all ASCII, e.g. lack of reliable support in all terminals.

**Comment by nnethercote** [comment]
> FWIW, deduplication could also have been accomplished by going the other direction

I have no opinions about that, so I'll let you decide whether to take my r=me, or ask someone with more diagnostic/span expertise for a second opinion.

File: tests/ui/inference/erase-type-params-in-label.stderr (hunk 2 above)
Comment (estebank) [reply tests/ui/inference/erase-type-params-in-label.stderr:5]: It's kind of blocked on having a `rustc.toml` file where we could customize output, so that people can change their configuration without much hassle when in less supported platforms. We should do this sooner rather than later. :-/
